File: Divide and Conquer/binarySearch.py
import math
import sys
import collections
import random
input=sys.stdin.readline
print=sys.stdout.write
sys.setrecursionlimit(1000000)

'''

Time Complexity: O(logN), N -> Number of elements
Space Complexity: O(N), N -> Number of elements

'''

def lin():
    return list(map(int,input().split()))

def binarySearch(arr,start,end,s):
    if start>end:
        return -1
    mid=(start+end)//2
    if arr[mid]==s:
        return mid
    if arr[mid]>s:
        return binarySearch(arr,start,mid-1,s)
    else:
        return binarySearch(arr,mid+1,end,s)
    return -1
 
if __name__=='__main__':
    n=int(input())
    arr=lin()
    s=int(input())
    result=binarySearch(arr,0,n-1,s)
    if result+1:
        print("Found at position : "+str(result+1)+'\n')
    else:
        print("Not found\n")
    

File: Divide and Conquer/binarySearch.cpp
#include<iostream>
#include<vector>
#include<math.h>
#include<iomanip>
#include<algorithm>
#include<string>
#include<map>
#include<bitset>
#include<cmath>
#include<iomanip>
#define vint vector<int>
#define ll long long
#define mod 1000000007
#define vll vector<long long>
#define ld long double
#define mii map<int,int>
#define miv map<int,vector<int>>
#define vf vector<float>
#define vd vector<double>

/*

Time Complexity: O(logN), N -> Number of elements
Space Complexity: O(N), N -> Number of elements

*/

using namespace std;

void solve()
{
	//Input
	int n,s;
	cin>>n;
	vint arr(n);
	for(int i=0;i<n;i++)
	{
		cin>>arr[i];
	}
	cin>>s;

	//Binary Search Algorithm
	int start=0,end=n-1,flag=-1;
	while(start<=end)
	{
		int mid=(start+end)/2;
		if(arr[mid]==s) // Searching middle element
		{
			flag=mid;
			break;
		}
		if(arr[mid]<s) // Moving to right sub-array
		{
			start=mid+1;
		}
		else // Moving to left sub-array
		{
			end=mid-1;
		}
	}
	if(flag+1)
	{
		cout<<"Found at position : "<<flag+1<<endl;
	}
	else
	{
		cout<<"Not found\n";
	}
	return;
}

int main()
{
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);
/*
	#ifndef ONLINE_JUDGE
	freopen("input.txt","r",stdin);
	freopen("output.txt","w",stdout);
	#endif
*/
	int t=1;
	while(t--)
	{
		solve();
	}
	return 0;
}
